js/testapps/vertexai/firestore/src/index.ts

```diff
@@ -0,0 +1,191 @@
+//  Sample app for using the proposed Vertex AI plugin retriever and indexer with Firestore.
+
+//  NOTE: This particular uses Firestore as the document store, but the plugin is unopinionated about the document store.
+
+import {
+  Document,
+  DocumentDataSchema,
+  index,
+  retrieve,
+} from '@genkit-ai/ai/retriever';
+import { configureGenkit } from '@genkit-ai/core';
+import { defineFlow, startFlowsServer } from '@genkit-ai/flow';
+// important imports for this sample:
+import {
+  DocumentIndexer,
+  DocumentRetriever,
+  Neighbor,
+  vertexAI,
+  vertexAiIndexerRef,
+  vertexAiRetrieverRef,
+} from '@genkit-ai/vertexai';
+import { initializeApp } from 'firebase-admin/app';
+import { getFirestore } from 'firebase-admin/firestore';
+import * as z from 'zod';
+
+// Environment variables set with dotenv for simplicity of sample
+import {
+  FIRESTORE_COLLECTION,
+  LOCATION,
+  PROJECT_ID,
+  VECTOR_SEARCH_DEPLOYED_INDEX_ID,
+  VECTOR_SEARCH_INDEX_ENDPOINT_ID,
+  VECTOR_SEARCH_INDEX_ID,
+  VECTOR_SEARCH_PUBLIC_ENDPOINT,
+} from './config';
+
+if (
+  [
+    FIRESTORE_COLLECTION,
+    LOCATION,
+    PROJECT_ID,
+    VECTOR_SEARCH_DEPLOYED_INDEX_ID,
+    VECTOR_SEARCH_INDEX_ENDPOINT_ID,
+    VECTOR_SEARCH_INDEX_ID,
+    VECTOR_SEARCH_PUBLIC_ENDPOINT,
+  ].some((envVar) => !envVar)
+) {
+  throw new Error(
+    'Missing environment variables. Please check your .env file.'
+  );
+}
+
+// Initialize Firebase app
+initializeApp({ projectId: PROJECT_ID });
+
+const db = getFirestore();
+
+// Firestore Document Retriever
+const firestoreRetriever: DocumentRetriever = async (
+  neighbors: Neighbor[]
+): Promise<Document[]> => {
+  const docs: Document[] = [];
+  for (const neighbor of neighbors) {
+    const docRef = db
+      .collection(FIRESTORE_COLLECTION)
+      .doc(neighbor.datapoint?.datapointId!);
+    const docSnapshot = await docRef.get();
+
+    // If this retriever fails to retrieve a document from a given id, it will not be included in the results.
+    // We could have included a placeholder document with an error message, but for simplicity we are just ignoring it.
+    // We could change it so DocumentRetriever returns a Promise.allSettled instead. The current behavior keeps types simpler though.
+    if (docSnapshot.exists) {
+      const docData = { ...docSnapshot.data(), metadata: { ...neighbor } };
+      const parsedDocData = DocumentDataSchema.safeParse(docData);
+      if (parsedDocData.success) {
+        docs.push(new Document(parsedDocData.data));
+      }
+    }
+  }
+  return docs;
+};
+
+// Firestore Document Indexer
+const firestoreIndexer: DocumentIndexer = async (
+  docs: Document[]
+): Promise<string[]> => {
+  const batch = db.batch();
+  const ids: string[] = [];
+  docs.forEach((doc) => {
+    const docRef = db.collection(FIRESTORE_COLLECTION).doc();
+    batch.set(docRef, { content: doc.content });
+    ids.push(docRef.id);
+  });
+  await batch.commit();
+  return ids;
+};
+
```

Comment:
Cool samples! I think there is value in providing a basic version of these implementations to help developers to get started with VVS. WDYT?

cc: @pavelgj @chrisraygill

Reply:
I agree - I think we should support Google databases out of the box, but make it easy for folks to integrate it with any other database

Reply:
Something like this? https://github.com/firebase/genkit/pull/519/files#diff-1243387b37033c5756d87d29782a2d1c0672ee3f8fdb2ed16a8fb56a8f2c8c26R22